src/hotspot/share/opto/phaseX.cpp

```diff
@@ -1991,24 +1991,35 @@ void PhaseCCP::push_load_barrier(Unique_Node_List& worklist, const BarrierSetC2*
   }
 }
 
-// AndI/L::Value() optimizes patterns similar to (v << 2) & 3 to zero if they are bitwise disjoint.
-// Add the AndI/L nodes back to the worklist to re-apply Value() in case the shift value changed.
-// Pattern: parent -> LShift (use) -> (ConstraintCast | ConvI2L)* -> And
+// AndI/L::Value() optimizes patterns similar to (v << 2) & 3, or CON & 3 to zero if they are bitwise disjoint.
+// Add the AndI/L nodes back to the worklist to re-apply Value() in case the value is now a constant or shift
+// value changed.
 void PhaseCCP::push_and(Unique_Node_List& worklist, const Node* parent, const Node* use) const {
+  const Node* to_push = nullptr;
+
+  const TypeInteger* new_type = type(parent)->isa_integer(type(parent)->basic_type());
   uint use_op = use->Opcode();
-  if ((use_op == Op_LShiftI || use_op == Op_LShiftL)
-      && use->in(2) == parent) { // is shift value (right-hand side of LShift)
-    auto push_and_uses_to_worklist = [&](Node* n){
-      uint opc = n->Opcode();
-      if (opc == Op_AndI || opc == Op_AndL) {
-        push_if_not_bottom_type(worklist, n);
-      }
-    };
-    auto is_boundary = [](Node* n) {
-      return !(n->is_ConstraintCast() || n->Opcode() == Op_ConvI2L);
-    };
-    use->visit_uses(push_and_uses_to_worklist, is_boundary);
+  if (new_type != nullptr && new_type->is_con()) {
+    // Pattern: parent (now constant) -> (ConstraintCast | ConvI2L)* -> And
+    to_push = parent;
+  } else if ((use_op == Op_LShiftI || use_op == Op_LShiftL) && use->in(2) == parent) {
+    // Pattern: parent -> LShift (use) -> (ConstraintCast | ConvI2L)* -> And
+    to_push = use;
+  }
+  if (to_push == nullptr) {
+    return;
   }
+
+  auto is_boundary = [](Node* n) {
+    return !(n->is_ConstraintCast() || n->Opcode() == Op_ConvI2L);
+  };
+  auto push_and_uses_to_worklist = [&](Node* n){
+    uint opc = n->Opcode();
+    if (opc == Op_AndI || opc == Op_AndL) {
+      push_if_not_bottom_type(worklist, n);
+    }
+  };
+  to_push->visit_uses(push_and_uses_to_worklist, is_boundary);
```

Comment:
And why not just call this line in two places, rather than having to work with `to_push`? Would that not be less code?

Reply:
Matthias reports:

That was purely a stylistic choice - the two cases can be integrated, we can push from "use" in both cases. Updated.